---
title: "RunMe_SingleCell.Rmd"
author: "Lucie Chrastecka"
date: "2023-03-24"
output: html_document
editor_options: 
  chunk_output_type: console
---

#This script will run the entirety of the appropriate steps of GeneSetMatch pipeline for Differential Expression Analysis (FindAllMarkers, Seurat, NMF) of a mixed mouse Liver, Lung, Brain single-cell RNA-Seq dataset. Pathway Enrichment is performed against C5 and C8  MSigDB sets, and visualized with our multimodal heatmaps. The heatmaps are very large, and therefore downloading separately from within the knitted html is recomended to preserve resolution and detail.
Assuming all necessary libraries are loaded, the computation should take 10-15 minutes.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(biomaRt)
library(base)
library(pdist)
library(gplots)
library(msigdbr)
library(GeneSetMatch)
library(BBmisc) #for normalize
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(xfun)
xfun::pkg_load2(c("htmltools", "mime"))
```


Make a matrix of sc LLB Tabula Muris expression Normalized
```{r}
TBM <- GeneSetMatch::TabulaMuris_LLB
TBM <- TBM$ex
TBM$geneID <- GeneSetMatch::MouseSYM2MouseENT(rownames(TBM))
rownames(TBM) <- NULL
TBM <- na.omit(TBM)
TBM$ENTREZID <- TBM$geneID
TBM <- TBM[,-4]
TBM_norm <- BBmisc::normalize(TBM, method = "range", range = c(0, 1), margin = 1L, on.constant = "quiet")#normalize this to be between 0-1, always must have 1 and 0 in each row
head(TBM_norm)

```

Make a matrix of sc LLB Tabula Muris expression FAM Max0
```{r}
TBM <- GeneSetMatch::TabulaMuris_LLB
Liver <- as.data.frame(TBM$GSEA$max0$hepactocyte$orig_matrix$stat, TBM$GSEA$max0$hepactocyte$orig_matrix$ENTREZID)
Liver$ENTREZID <- rownames(Liver)
Liver$Liver <- Liver$`TBM$GSEA$max0$hepactocyte$orig_matrix$stat`
Liver$`TBM$GSEA$max0$hepactocyte$orig_matrix$stat` <- NULL
Liver <- Liver[!is.na(Liver$ENTREZID),]
head(Liver)
dim(Liver)
#for each gene that appears more than one time, take the average and keep only one instance
#remove na's

Lung <- as.data.frame(TBM$GSEA$max0$lung$orig_matrix$stat, TBM$GSEA$max0$lung$orig_matrix$ENTREZID)
Lung$ENTREZID <- rownames(Lung)
Lung$Lung <- Lung$`TBM$GSEA$max0$lung$orig_matrix$stat`
Lung$`TBM$GSEA$max0$lung$orig_matrix$stat` <- NULL
Lung <- Lung[!is.na(Lung$ENTREZID),]
head(Lung)
dim(Lung)

Brain <- as.data.frame(TBM$GSEA$max0$microglia$orig_matrix$stat, TBM$GSEA$max0$microglia$orig_matrix$ENTREZID)
Brain$ENTREZID <- rownames(Brain)
Brain$Brain <- Brain$`TBM$GSEA$max0$microglia$orig_matrix$stat`
Brain$`TBM$GSEA$max0$microglia$orig_matrix$stat` <- NULL
Brain <- Brain[!is.na(Brain$ENTREZID),]
head(Brain)
dim(Brain)


TBM_fam <- merge(Liver, Lung, by = "ENTREZID", all = FALSE)
TBM_fam <- merge(TBM_fam, Brain, by = "ENTREZID", all = FALSE)
#TBM_fam <- merge(TBM_fam, Brain)
head(TBM_fam)
dim(TBM_fam)

#MEGA MERGE, TO MAKE SURE BOTH MATRIXES ARE SAME SIZE
TBM_all <- merge(TBM_norm, TBM_fam, by = "ENTREZID", all = FALSE)
head(TBM_all)
dim(TBM_all)
```


Preparing MSigDB Sets and conversions into .gmt files for GSEA
```{r}
C5_df = msigdbr(species = "Mus musculus", category = "C5", subcategory = "BP")
C8_df = msigdbr(species = "Mus musculus", category = "C8", subcategory = "") #update on all

list_of_sets = list(C5 = C5_df,
                    C8 = C8_df)

list_of_gmt_files <- GeneSetMatch::convert_msigdbr_obj_to_gmt_file(list_of_sets)
```

#GSEA vectors
```{r}
TBM_FAM <- TBM_all[,c(1,5:7)]
TBM_FAM <- data.frame(TBM_FAM, row.names = TBM_FAM[,1])
TBM_FAM <- TBM_FAM[,-1]
head(TBM_FAM)
TBM_FAM <- as.matrix(TBM_FAM)
head(TBM_FAM)
typeof(TBM_FAM) #SHOULD BE "DOUBLE"


TBM_NORM <- TBM_all[,1:4]
TBM_NORM <- data.frame(TBM_NORM, row.names = TBM_NORM[,1])
TBM_NORM <- TBM_NORM[,-1]
head(TBM_NORM)
TBM_NORM <- as.matrix(TBM_NORM)
head(TBM_NORM)
typeof(TBM_NORM) #SHOULD BE "DOUBLE"

```

GSEA on all three vectors of NMF +orig values
```{r echo=TRUE, results='hide'}

TBM_GSEA <- list()

## Both of the following lists have naming convention "v1","v2", etc, so you can index them in a single for loop by assigning the loop variable as the character string.

## Make list of norm aka raw scores
Q = list(V1 = TBM_NORM[,1, drop = F],
         V2 = TBM_NORM[,2, drop = F],
         V3 = TBM_NORM[,3, drop = F])


## make list of tfpm scores.
new_y = list(V1 = TBM_FAM[,1, drop = F],
             V2 = TBM_FAM[,2, drop = F],
             V3 = TBM_FAM[,3, drop = F])
    
    
for(Y in c("V1","V2","V3") #drop = f MEANS KEEP IT AS A MATRIX WITH COLLUMNS AND
){
  
  ## Make vector sorted gene list of TFPM scores
    X <- new_y[[Y]]
    gene_list = X[,1]
    gene_list = sort(gene_list, decreasing = TRUE)
    gene_list = gene_list[!duplicated(names(gene_list))]
    print("head gene_list")
    head(gene_list, n = 5)
    
    ## Make vector sorted gene list of raw nmf scores
    Z <- Q[[Y]]
    orig_value = Z[,1]
    orig_value = orig_value[!duplicated(names(orig_value))]
    orig_value <- orig_value[match(names(gene_list), names(orig_value))] #SORTED BASED ON GENE_LIST TO MAINTAIN ORDER
    print("head orig_value")
    head(orig_value, n = 5)
    
    head(names(gene_list),n=5)
    
    print(" Are these three things ^^ sorted the same way ??!?!?!?!?!?")
    
    
    for(i in list_of_gmt_files){
      print(i)
      res <- GeneSetMatch::GSMGSEA(gene_list = gene_list, theGoFile = i, pval = 0.1, onlypos = TRUE) #drop to 0.1, onlypos is true because NMF
      TBM_GSEA[[Y]][[i]] <- res  ## find a way to reference "nmf_norm" and "nmf_norm_org"
    }
    #orig_matrix must be a data frame
    TBM_GSEA[[Y]]$orig_matrix <-
      as.data.frame(
        list(
          log2FoldChange = orig_value,
          EnrichmentScore = gene_list,
          ENTREZID = names(gene_list),
          SYMBOL = GeneSetMatch::MouseENT2MouseSYM(names(gene_list))
        )
      ) #rat conversion function returns some NA's 
    
}

summary(TBM_GSEA)

```

#DE Heatmaps for all .gmt files
```{r echo=TRUE, results='hide'}
  GeneSetMatch::GSM.Multilevel.Mouse.Heatmaps(TBM_GSEA, verbose = FALSE, pdftofile = TRUE, filePrefix = "scRNA")
  
```

#Mutlimodal Heatmap for C5 geneset (Very Large)
```{r}
xfun::embed_file('./myC5.gmt.scRNA.pdf')
```

#multimodal Heatmap for C8 geneset (Very Large)
```{r}
xfun::embed_file('./myC8.gmt.scRNA.pdf')
```
